packages/flutter/lib/src/widgets/resolution_awareness.dart

```diff
@@ -0,0 +1,127 @@
+///
+/// heart.meta
+/// 1.0x/heart.png
+/// 1.5x/heart.png
+/// 2.0x/heart.png
+///
+/// On a device with a 1.0 device pixel ratio, the image chosen would be
+/// 1.0x/heart.png; on a device with a 1.3 device pixel ratio, the image chosen
+/// would be 1.5x/heart.png.
+///
+/// Common prefixes will be removed, so the following would be a valid bundle
+/// layout for an asset group named "icons/heart":
+///
+/// icons/heart.meta
+/// icons/1.0x/heart.png
+/// icons/1.5x/heart.png
+/// icons/2.0x/heart.png
+class ResolutionAwareAssetResolver extends CandidateAssetResolver {
+  ResolutionAwareAssetResolver({ AssetBundle bundle, this.devicePixelRatio })
+    : super(bundle: bundle);
+
+  final double devicePixelRatio;
+
+  SplayTreeMap<double, String> _buildMapping(List<String> candidates) {
+    SplayTreeMap<double, String> result = new SplayTreeMap<double, String>();
+    RegExp exp = new RegExp(r"/?(\d+(\.\d*)?)x/");
+    for (String candidate in candidates) {
+      Match match = exp.firstMatch(candidate);
+      if (match != null && match.groupCount > 0) {
+        double resolution = double.parse(match.group(1));
+        result[resolution] = candidate;
+      }
+    }
+    return result;
+  }
+
+  String chooseCandidate(List<String> candidates) {
+    return new BreakpointQualifier(
+      candidates: _buildMapping(candidates),
+      value: devicePixelRatio
+    ).resolve();
+  }
+}
+
+/// Configuration for a ResolutionAwareAssetResolver.
+class ResolutionAwareResolverConfig extends AssetResolverConfig {
+  const ResolutionAwareResolverConfig({ this.devicePixelRatio, this.bundle });
+
+  final double devicePixelRatio;
+  final AssetBundle bundle;
+
+  ResolutionAwareAssetResolver createResolver() =>
+    new ResolutionAwareAssetResolver(
+      bundle: bundle,
+      devicePixelRatio: devicePixelRatio
+    );
+
+  bool operator==(ResolutionAwareResolverConfig config) =>
+    devicePixelRatio == config.devicePixelRatio &&
+    bundle == config.bundle;
+
+  int get hashCode => devicePixelRatio.hashCode + bundle.hashCode;
+}
+
+/// Displays an image from an [AssetBundle] using the
+/// [ResolutionAwareAssetResolver] resolution strategy.
+class ResolutionAwareAssetImage extends AssetImage {
+  ResolutionAwareAssetImage({
+    Key key,
+    String name,
+    AssetBundle bundle,
+    double width,
+    double height,
+    ColorFilter colorFilter,
+    ImageFit fit,
+    FractionalOffset alignment,
+    ImageRepeat repeat: ImageRepeat.noRepeat,
+    Rect centerSlice
+  }) : super(
+    key: key,
+    name: name,
+    bundle: bundle,
+    width: width,
+    height: height,
+    colorFilter: colorFilter,
+    fit: fit,
+    alignment: alignment,
+    repeat: repeat,
+    centerSlice: centerSlice
+  );
+
+  ImageResource loadImage(BuildContext context) {
+    AssetBundle currentBundle = bundle ?? DefaultAssetBundle.of(context);
+    ResolutionAwareResolverConfig config = new ResolutionAwareResolverConfig(
```

Comment:
"config" is a member with meaning in State classes, and code from Widget classes is often copied to State classes, so we should probably steer away from that name here.

As a side note, in general, we try to avoid abbreviating names. ("config" on State is an intentional exception).